In [ ]:
# Install the required packages
!pip3 install -U elasticsearch
!pip3 install -U jsonlines

     |████████████████████████████████| 225kB 2.7MB/s 


In [ ]:
from datetime import datetime
from elasticsearch import Elasticsearch
from elasticsearch import helpers
import time
import jsonlines
import pandas as pd

In [ ]:
# you can use RFC-1738 to specify the url | Initialize the ES client 
host = "https://search-covidtweets-pc7dqt6roqlapjsdifs6fk6oy4.us-east-1.es.amazonaws.com/"

es = Elasticsearch(['https://search-covidtweets-pc7dqt6roqlapjsdifs6fk6oy4.us-east-1.es.amazonaws.com/'], sniff_on_connection_fail= True) #, timeout=500000) << Not certain these parameters add value

In [ ]:
# Specify the search query via the json & execute the command 
# This is sample to create a baseline 
body = {
    "from": 0, # index from which the search is carried out
    "size":100, # Number of matches that should be returned
    "_source": ["created_at", "user.screen_name","user.location", "full_text"],
    "query" : {
        "constant_score" : {
            "filter" : {
                "bool": {
                    "must_not": {"term": {"user.location": ''}}
                }
            }
        }
    
}
}

q1 = es.search(index='covidtweets*', body=body)


In [ ]:
pd.DataFrame(q1['hits']['hits'])

_index  ...                                            _source
0   covidtweets  ...  {'full_text': 'RT @wsfa12news: UPDATE: Alabama...
1   covidtweets  ...  {'full_text': '@DmKasganj @CMOfficeUP जय हिंद ...
2   covidtweets  ...  {'full_text': 'RT @meiccymru: Already bored of...
3   covidtweets  ...  {'full_text': 'RT @AyinAzirinee: Good article ...
4   covidtweets  ...  {'full_text': 'RT @SupJvse: What Latino parent...
..          ...  ...                                                ...
95  covidtweets  ...  {'full_text': 'RT @CNN: This is how the Wuhan ...
96  covidtweets  ...  {'full_text': 'RT @Doranimated: US and British...
97  covidtweets  ...  {'full_text': 'RT @aravosis: Yep.', 'created_a...
98  covidtweets  ...  {'full_text': 'RT @joncoopertweets: The corona...
99  covidtweets  ...  {'full_text': '🔴 #irdirc needs u! We would lik...

[100 rows x 5 columns]

In [ ]:
# Specify the search query via the json & execute the command
body = {
    "from": 0, # index from which the search is carried out
    "size":1000, # Number of matches that should be returned
    "_source": ["created_at", "user.screen_name","user.location", "user.description", "full_text"],
    "query": {
    "query_string": {
      "query": "(advocaat) OR (staatssecretaris) OR (manager) OR (gezondheidscentrum) OR (medisch) OR (welzijnsinstelling) OR (agent) OR (apoth)"
      
    }
  }
    
      }

q1 = es.search(index='dutch*', body=body)
query_DF = pd.DataFrame(q1['hits']['hits'])

++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

The next few code blocks I will try to implement a scroll API to extract all the docs from an index 

Source:
https://kb.objectrocket.com/elasticsearch/how-to-use-python-to-make-scroll-queries-to-get-all-documents-in-an-elasticsearch-index-752

In [ ]:
# import the Elasticsearch client library
from elasticsearch import Elasticsearch, exceptions

# import JSON and time
import json, time

index = "dutch*"

In [ ]:
# declare a filter query dict object
match_all = {
    "size": 5000,
    "_source": ["created_at", "user.screen_name","user.location", "user.description", "full_text"],
    "query": {
        "match_all": {}
        }
    }

# make a search() request to get all docs in the index
resp = es.search(
            index = 'dutch*',
            body = match_all,
            scroll = '15s' # length of time to keep search context
        )

# keep track of pass scroll _id
old_scroll_id = resp['_scroll_id']

In [ ]:
allData = pd.DataFrame()

In [ ]:
azaaaaaAAQA
time.sleep(5)

# use a 'while' iterator to loop over document 'hits'
while len(resp['hits']['hits']):
            query_DF = pd.DataFrame(resp['hits']['hits'])
            queryDF = pd.DataFrame(list(query_DF['_source']))
            user_info = queryDF['user'].apply(pd.Series)
            queryDF = pd.concat( [queryDF.drop(['user'],axis=1) , user_info], axis=1)
            allData = pd.concat([allData, queryDF], axis =0)
            # make a request using the Scroll API
            resp = es.scroll(
                scroll_id = old_scroll_id,
                scroll = '15s' # length of time to keep search context
            )

            # check if there's a new scroll ID
            if old_scroll_id != resp['_scroll_id']:
                print ("NEW SCROLL ID:", resp['_scroll_id'])

            # keep track of pass scroll _id
            old_scroll_id = resp['_scroll_id']

            # print the response results
            print ("\nresponse for index:", index)
            print ("_scroll_id:", resp['_scroll_id'])
            print ('Collected data:', str(allData.shape[0]))
            time.sleep(5)
            

NotFoundError: ignored

++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

In [ ]:
allData.shape

(201721, 5)

In [ ]:
allData.head()

full_text  ...                    location
0  @pflegearzt @Friedelkorn @LAguja44 Pardon, wol...  ...                 Netherlands
1  RT @grantshapps: Aviation demand is reduced du...  ...  Amsterdam, The Netherlands
2  RT @DDStandaard: De droom van D66 wordt werkel...  ...                            
3  De droom van D66 wordt werkelijkheid: COVID-19...  ...                   Amsterdam
4  RT @eucopresident: Following consultations I w...  ...                   Amsterdam

[5 rows x 5 columns]

In [ ]:
queryDF = pd.DataFrame(list(query_DF['_source']))

In [ ]:
user_info = queryDF['user'].apply(pd.Series)

In [ ]:
queryDF = pd.concat( [queryDF.drop(['user'],axis=1) , user_info], axis=1)

In [ ]:
queryDF.head(15)

full_text  ...                    location
0   @pflegearzt @Friedelkorn @LAguja44 Pardon, wol...  ...                 Netherlands
1   RT @grantshapps: Aviation demand is reduced du...  ...  Amsterdam, The Netherlands
2   RT @DDStandaard: De droom van D66 wordt werkel...  ...                            
3   De droom van D66 wordt werkelijkheid: COVID-19...  ...                   Amsterdam
4   RT @eucopresident: Following consultations I w...  ...                   Amsterdam
5       Read absorb and share https://t.co/IHPDY6qS4w  ...             The Netherlands
6   RT @FrankvanWijck: Het aantal deskundigen over...  ...                            
7   RT @HanTenBroeke: Ook zo benieuwd of het Coron...  ...                     NLD-DEU
8   Het wordt dommer en dommer! We weten nu alleen...  ...  Amsterdam, The Netherlands
9   @volkskrant We hebben toch niet een ander viru...  ...  Amsterdam, The Netherlands
10  Ook hier komt het #coronavirus helaas steeds d...  ...                 Netherlands
11  RT @spectatorindex: Stockmarkets today.\n\nIta...  ...                 Netherlands
12  Corona virus is sws gemaakt door handzeep bedr...  ...                            
13  @KHJ_Esq The Corona-crisis offers great opport...  ...             The Netherlands
14  RT @parool: Het coronavirus breidt zich verder...  ...             Lopik / Holland

[15 rows x 5 columns]

In [ ]:
queryDF.tail(15)

full_text  ...                     location
985  Steeds meer bedrijven blunderen met hun corona...  ...       The Hague, Netherlands
986            Corona test came in and it's negative 🥳  ...        Nijmegen, Netherlands
987  Everyone please wear a damn mask so we can go ...  ...  Amsterdam, The Netherlands 
988  RT @jorisvancauter: Goed dat het parlement op ...  ...                             
989  RT @frankvanardenne: “Tussen 17 maart tot 21 m...  ...                             
990  @VogelvrijeHArts Er is bij zes medewerkers van...  ...                     Universe
991  RT @RutgervdNoort: Leestip:” Medisch gezien is...  ...  Nederland / The Netherlands
992  RT @LudwigBollaerts: Zullen we deze foto morge...  ...                             
993  RT @EricChrSmit: Eelco van Anken, celbioloog e...  ...               Usan, Scotland
994  RT @Bloempje141: Deze meneer met verstand en e...  ...                             
995  RT @ZZPLoosduinen: "Ik ben medisch moleculair ...  ...                    Nederland
996  RT @ZZPLoosduinen: "Ik ben medisch moleculair ...  ...                             
997  RT @biancadenouter: ‘Laten we een beetje nucht...  ...        Westerkwartier, Grins
998  RT @fvdemocratie: #FVD-senator @NickiVerweij l...  ...                    Amsterdam
999  @Ella_Ster huisarts kan wel voorschrijven, maa...  ...                             

[15 rows x 5 columns]

In [ ]:
allData.to_csv('dutch_tweets.csv', index=False)

In [ ]:
!pip3 install -U spacy

In [ ]:
!python3 -m spacy download nl_core_news_sm

In [ ]:
import spacy

In [ ]:
nlp = spacy.load("nl_core_news_sm")

In [ ]:
doc = nlp("Dit is een zin.")

In [ ]:
print([(w.text, w.pos_) for w in doc])

[('Dit', 'PRON'), ('is', 'VERB'), ('een', 'DET'), ('zin', 'NOUN'), ('.', 'SYM')]


In [ ]:
for ent in doc.ents:
    print(ent.text, ent.start_char, ent.end_char, ent.label_)

In [ ]:
for token in doc:
    print(token.text, token.lemma_, token.pos_, token.tag_, token.dep_,
            token.shape_, token.is_alpha, token.is_stop)

Dit dit PRON VNW|aanw|pron|stan|vol|3o|ev__Person=3|PronType=Dem nsubj Xxx True True
is zijn VERB WW|pv|tgw|ev__Number=Sing|Tense=Pres|VerbForm=Fin cop xx True True
een een DET LID|onbep|stan|agr__Definite=Ind det xxx True True
zin zin NOUN N|soort|ev|basis|zijd|stan__Gender=Com|Number=Sing ROOT xxx True False
. . SYM LET punct . False False


In [ ]:
for chunk in doc.noun_chunks:
    print(chunk.text, chunk.root.text, chunk.root.dep_,
            chunk.root.head.text)

In [ ]:
def checkWords(textString):
  matches = ['advocaat', 'staatssecretaris', 'gezondheidscentrum', 'welzijnsinstelling', 'medisch']

  if any(x in textString for x in matches):
    return True
  else:
    return False


In [ ]:
queryDF['Industry'] = queryDF['full_text'].apply(checkWords)

In [ ]:
sum(queryDF['Industry'])

167

In [ ]:
queryDF.head(15)

full_text  ... Industry
0   RT @NVALeuven: Onaanvaardbaar! 😡😡😡\n\nSpoedig ...  ...    False
1   Coronabonus voor leiding Zaans Medisch Centrum...  ...    False
2   Dubbele verantwoordelijkheid dus voor werkgeve...  ...    False
3   RT @Elsdijkstra95: Nee tatuurlijk niet he. Waa...  ...    False
4   RT @mikercarpenter: News that got completely b...  ...    False
5   RT @FaQuix: Nog zo’n willekeur, de kleurencode...  ...    False
6   RT @RubenKorthorst: De ‘willing to pay’ lijkt ...  ...    False
7   Zij die vinden dat covid19 een voldoende concr...  ...    False
8   RT @FleurAgemaPVV: Ingezet voor de zorg? Je be...  ...    False
9   RT @bslagter: (A) Van Dissel legt in die video...  ...    False
10  RT @Museumschat: Ik vind het zo erg dat hier w...  ...    False
11  RT @HrothN: Wat u niet zult zien bij de staats...  ...    False
12  RT @fransweisglas: Terechte woede. Maar dit is...  ...    False
13  Nederland gaat verder open. Coronavirus heeft ...  ...     True
14  Onbegrijpelijk dat de EU de regels voor het te...  ...     True

[15 rows x 6 columns]

In [ ]:
queryDF.to_excel('query.xlsx', index=False)

++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

This code block will calculate the weekday/ month/ year dat points 

In [ ]:
import datetime

allData['dateObject'] = allData['created_at'].apply(lambda x: datetime.datetime.strptime(x, '%a %b %d %H:%M:%S %z %Y'))

allData['Day'] = allData['dateObject'].apply(lambda x: x.day)
allData['Weekday'] = allData['dateObject'].apply(lambda x: x.weekday)
allData['Month'] = allData['dateObject'].apply(lambda x: x.month)
allData['Year'] = allData['dateObject'].apply(lambda x: x.year)
allData.drop(['dateObject'], axis=1, inplace=True)

In [ ]:
allData.head()

full_text  ... Day
0  @pflegearzt @Friedelkorn @LAguja44 Pardon, wol...  ...   9
1  RT @grantshapps: Aviation demand is reduced du...  ...   9
2  RT @DDStandaard: De droom van D66 wordt werkel...  ...   9
3  De droom van D66 wordt werkelijkheid: COVID-19...  ...   9
4  RT @eucopresident: Following consultations I w...  ...   9

[5 rows x 9 columns]

++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

The next code block will use the geoCoder API (OSM) to get the lat,long for each of the location (unique) in the dataset

In [ ]:
import geopy.geocoders
from geopy.geocoders import Nominatim
from geopy.extra.rate_limiter import RateLimiter


In [ ]:
locator = Nominatim(user_agent="myGeocoder")

# 1 - conveneint function to delay between geocoding calls
geocode = RateLimiter(locator.geocode, min_delay_seconds=1)


In [ ]:
# Get all unique locations 

allLocation = allData['location'].unique().tolist()
print("Nuber of unique locations",len(allLocation))

Nuber of unique locations 15283


In [ ]:
allLocation =pd.DataFrame(allLocation)
print(allLocation.shape)
allLocation.columns = ['location']
print(allLocation.head())

(15283, 1)
                     location
0                 Netherlands
1  Amsterdam, The Netherlands
2                            
3                   Amsterdam
4             The Netherlands


I have a one second rate limiter so for 15283 records it may take a lot of time! 42hrs+ 

So lets try to break up the records into smaller units & run them in parallel

In [ ]:
import numpy as np

allLocationArray = np.array_split(allLocation, 10)
allLocationArray[1]

location
1529                         Veere
1530             overal en nergens
1531  Earth,The One and Only Truth
1532       Diepenveen, Netherlands
1533                    Oosterhout
...                            ...
3053   Leiden, Den Haag, Teylingen
3054        Liskwartier, Rotterdam
3055               Scheveningen ..
3056             Het Heuge Neurden
3057            Zutphen, Nederland

[1529 rows x 1 columns]

In [ ]:
for idx in range(len(allLocationArray)):
  allLocationArray[idx].to_excel("allLocation_chunk"+str(idx)+".xlsx", index=True)

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
!cp drive/"My Drive"/"Dutch Tweets/allLocation_chunk8.xlsx" .

cp: cannot stat 'drive/My Drive/Dutch Tweets/allLocation_chunk8.xlsx': No such file or directory


In [ ]:
!ls drive/"My Drive"/"Dutch Tweets"

allLocation_chunk0.csv	allLocation_chunk4.csv	allLocation_chunk8.csv
allLocation_chunk1.csv	allLocation_chunk5.csv	allLocation_chunk9.csv
allLocation_chunk2.csv	allLocation_chunk6.csv	dutch_tweets.csv
allLocation_chunk3.csv	allLocation_chunk7.csv


In [ ]:
drive.flush_and_unmount()
print('All changes made in this colab session should now be visible in Drive.')

All changes made in this colab session should now be visible in Drive.


In [ ]:
allDataList = np.array_split(allData, 10)

In [ ]:
for idx in range(len(allDataList)):
  print(idx)
  try:
    allDataList[idx].to_csv("allDataList_chunk"+str(idx)+".csv", index=True)
  except:
    print("ERR in idx: ", idx)

0
1
2
3
4
5
6
7
8
9


In [ ]:
!cp allDataList*.csv drive/"My Drive"/"Dutch Tweets"
!ls drive/"My Drive"/"Dutch Tweets"

allDataList_chunk0.csv	allDataList_chunk8.csv	   allLocation_chunk5.csv
allDataList_chunk1.csv	allDataList_chunk9.csv	   allLocation_chunk6.csv
allDataList_chunk2.csv	allDataList_chunk9.gsheet  allLocation_chunk7.csv
allDataList_chunk3.csv	allLocation_chunk0.csv	   allLocation_chunk8.csv
allDataList_chunk4.csv	allLocation_chunk1.csv	   allLocation_chunk9.csv
allDataList_chunk5.csv	allLocation_chunk2.csv	   dutch_tweets.csv
allDataList_chunk6.csv	allLocation_chunk3.csv
allDataList_chunk7.csv	allLocation_chunk4.csv


In [ ]:
drive.flush_and_unmount()
print('All changes made in this colab session should now be visible in Drive.')

All changes made in this colab session should now be visible in Drive.


In [ ]:
# Read location chunks 
import pandas as pd 

idx = 8

allLocation_chunk = pd.read_excel("allLocation_chunk"+str(idx)+".xlsx")
allLocation_chunk.shape
allLocation_chunk.head()

Unnamed: 0                        location
0       12227         Renkum, The Netherlands
1       12228   Noord-Holland en Zuid-Holland
2       12229            Sin city / Amsterdam
3       12230      Heeswijk-Dinther, Bernheze
4       12231  Hintham-Zuid, 's-Hertogenbosch

In [ ]:
# 2- - create location column
allLocation_chunk['point_info'] = allLocation_chunk['location'].apply(geocode)

RateLimiter caught an error, retrying (0/2 tries). Called with (*('N.v.t.',), **{}).
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/geopy/geocoders/base.py", line 344, in _call_geocoder
    page = requester(req, timeout=timeout, **kwargs)
  File "/usr/lib/python3.6/urllib/request.py", line 526, in open
    response = self._open(req, data)
  File "/usr/lib/python3.6/urllib/request.py", line 544, in _open
    '_open', req)
  File "/usr/lib/python3.6/urllib/request.py", line 504, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.6/urllib/request.py", line 1368, in https_open
    context=self._context, check_hostname=self._check_hostname)
  File "/usr/lib/python3.6/urllib/request.py", line 1328, in do_open
    r = h.getresponse()
  File "/usr/lib/python3.6/http/client.py", line 1356, in getresponse
    response.begin()
  File "/usr/lib/python3.6/http/client.py", line 307, in begin
    version, status, reason = self._read_status()
  File "/u

In [ ]:
# 3 - create longitude, laatitude and altitude from location column (returns tuple)
allLocation_chunk['point'] = allLocation_chunk['point_info'].apply(lambda loc: tuple(loc.point) if loc else None)

allLocation_chunk.head()

Unnamed: 0                        location  ... altitude       Province
0       12227         Renkum, The Netherlands  ...      0.0     Gelderland
1       12228   Noord-Holland en Zuid-Holland  ...      0.0   Zuid-Holland
2       12229            Sin city / Amsterdam  ...      0.0          False
3       12230      Heeswijk-Dinther, Bernheze  ...      0.0  Noord-Brabant
4       12231  Hintham-Zuid, 's-Hertogenbosch  ...      0.0  Noord-Brabant

[5 rows x 8 columns]

In [ ]:
# 4 - split point column into latitude, longitude and altitude columns
allLocation_chunk[['latitude', 'longitude', 'altitude']] = pd.DataFrame(allLocation_chunk['point'].tolist(), index=allLocation_chunk.index)

allLocation_chunk.head(15)

Unnamed: 0                        location  ... altitude             Province
0        12227         Renkum, The Netherlands  ...      0.0           Gelderland
1        12228   Noord-Holland en Zuid-Holland  ...      0.0         Zuid-Holland
2        12229            Sin city / Amsterdam  ...      0.0                False
3        12230      Heeswijk-Dinther, Bernheze  ...      0.0        Noord-Brabant
4        12231  Hintham-Zuid, 's-Hertogenbosch  ...      0.0        Noord-Brabant
5        12232           Grunobuurt, Groningen  ...      0.0            Groningen
6        12233             Lochem-Zuid, Lochem  ...      NaN                False
7        12234        ÜT: 36.098565,-80.311023  ...      NaN                False
8        12235          BTS |﹙G﹚I﹣ᴅʟᴇ | TOP 🇵🇹  ...      NaN                False
9        12236                     La Capitale  ...      0.0                False
10       12237                    Merksplas 🇧🇪  ...      0.0                False
11       12238           Driebergen, Nederland  ...      0.0              Utrecht
12       12239          Veenendaal | Apeldoorn  ...      NaN                False
13       12240                         Teheran  ...      0.0                False
14       12241                 Schiermonnikoog  ...      0.0  Friesland (Fryslân)

[15 rows x 8 columns]

In [ ]:
!cp drive/"My Drive"/"Dutch Tweets"/*.geojson .

In [ ]:
!pip install geopandas
!pip install geopy


In [ ]:
import geopandas as gpd

provinces = gpd.read_file('provinces.geojson')

print(provinces.shape)
provinces.head()

(12, 4)


name  ...                                           geometry
0              Drenthe  ...  POLYGON ((6.41328 52.98552, 6.36252 53.03397, ...
1            Flevoland  ...  POLYGON ((5.36115 52.67573, 5.37726 52.76481, ...
2  Friesland (Fryslân)  ...  POLYGON ((5.08707 53.32307, 5.10178 53.36803, ...
3           Gelderland  ...  POLYGON ((5.60602 51.99416, 5.59200 52.00139, ...
4            Groningen  ...  POLYGON ((6.28698 53.34138, 6.27368 53.34527, ...

[5 rows x 4 columns]

In [ ]:
from shapely.geometry import Point, Polygon

In [ ]:
def checkProvince(p1):
  #print(len(p1))
  if p1 != None and len(p1) >= 2 :
    p1 = Point(p1[1], p1[0])  
    provinces['InProvince'] = provinces['geometry'].apply(lambda x: x.contains(p1))
    selProvince = provinces[provinces['InProvince'] == True]
    #print(selProvince)
    if selProvince.shape[0] > 0:
      return selProvince['name'].values[0]
    else:
      return False
  else:
    return False

In [ ]:
allLocation_chunk['Province'] = allLocation_chunk['point'].apply(lambda x: checkProvince(x))
allLocation_chunk.head(15)

Unnamed: 0                        location  ... altitude             Province
0        12227         Renkum, The Netherlands  ...      0.0           Gelderland
1        12228   Noord-Holland en Zuid-Holland  ...      0.0         Zuid-Holland
2        12229            Sin city / Amsterdam  ...      0.0                False
3        12230      Heeswijk-Dinther, Bernheze  ...      0.0        Noord-Brabant
4        12231  Hintham-Zuid, 's-Hertogenbosch  ...      0.0        Noord-Brabant
5        12232           Grunobuurt, Groningen  ...      0.0            Groningen
6        12233             Lochem-Zuid, Lochem  ...      NaN                False
7        12234        ÜT: 36.098565,-80.311023  ...      NaN                False
8        12235          BTS |﹙G﹚I﹣ᴅʟᴇ | TOP 🇵🇹  ...      NaN                False
9        12236                     La Capitale  ...      0.0                False
10       12237                    Merksplas 🇧🇪  ...      0.0                False
11       12238           Driebergen, Nederland  ...      0.0              Utrecht
12       12239          Veenendaal | Apeldoorn  ...      NaN                False
13       12240                         Teheran  ...      0.0                False
14       12241                 Schiermonnikoog  ...      0.0  Friesland (Fryslân)

[15 rows x 8 columns]

In [ ]:
# idx = 0
allLocation_chunk.to_csv("allLocation_chunk"+str(idx)+".csv", index=False)
print("Geo processing completed....allLocation_chunk"+str(idx) + ".csv")

Geo processing completed....allLocation_chunk8.csv


In [ ]:
!cp allLocation_chunk8.csv drive/"My Drive"/"Dutch Tweets"/allLocation_chunk8.csv

In [ ]:
temp = pd.read_csv("allLocation_chunk8.csv")